File: kh2_src/tabs/journal_tabs/bestiary.py
from dash import Dash, html, dcc, callback, Input, Output, State, ALL
import kh2_src.kh2_utils as utils


@callback(
    Output("JournalBestiaryDiv", "children"),
    Input("JournalBestiaryTabs", "value"),
)
def __create_journal_characters_tabs(tab):
    if tab == "The Heartless":
        return __create_the_heartless()
    if tab == "The Nobodies":
        return __create_nobodies()
    if tab == "Reaction Commands":
        return __create_rcs()
    if tab == "Limits":
        return __create_limits()

def __create_the_heartless():
    kh2 = utils.kh2
    heartless = html.Div([
        html.Div([
            html.Label(k + ": "),
            dcc.Input(
                id={"type": "Heartless Kill Count", "index": kh2.heartless_dict[k]},
                type="number",
                value=kh2.heartless[kh2.heartless_dict[k]],
                min=0,
                max=0xFFFFFFFF,
                step=1,
                style={"width": 100},
            ),
        ],
            style={"margin-top": 20, "gap": 10},
        ) for k in kh2.heartless_list if kh2.heartless_dict[k] < len(kh2.heartless)
    ])
    return html.Div([
        heartless,
    ])

def __create_nobodies():
    kh2 = utils.kh2
    nobodies = html.Div([
        html.Div([
            html.Label(k + ": "),
            dcc.Input(
                id={"type": "Nobody Kill Count", "index": kh2.nobody_dict[k]},
                type="number",
                value=kh2.nobodies[kh2.nobody_dict[k]],
                min=0,
                max=0xFFFFFFFF,
                step=1,
                style={"width": 100},
            ),
        ],
            style={"margin-top": 20, "gap": 10},
        ) for k in kh2.nobody_list
    ])
    return html.Div([
        nobodies,
    ])

def __create_rcs():
    kh2 = utils.kh2
    rcs = html.Div([
        html.Div([
            html.H3(k + ":"),
            html.Div([
                html.Div([
                    html.Label(rc + ": "),
                    dcc.Input(
                        id={"type": "RC Count", "index": kh2.rc_dict[rc]},
                        type="number",
                        value=kh2.rc_usage[kh2.rc_dict[rc]],
                        min=0,
                        max=0xFFFFFFFF,
                        step=1,
                        style={"width": 100},
                    )
                ],
                    style={"margin-top": 20, "gap": 10},
                ) for rc in v
            ]),
        ]) for k, v in kh2.rc_list_dict.items()
    ])
    return html.Div([
        rcs,
    ])

def __create_limits():
    kh2 = utils.kh2
    limits = html.Div([
        html.Div([
            html.Label(k + ": "),
            dcc.Input(
                id={"type": "Limit Count", "index": kh2.limit_dict[k]},
                type="number",
                value=kh2.limit_usage[kh2.limit_dict[k]],
                min=0,
                max=0xFFFFFFFF,
                step=1,
                style={"width": 100},
            ),
        ],
            style={"margin-top": 20, "gap": 10},
        ) for k in kh2.limit_list
    ])
    return html.Div([
        limits,
    ])


def create_bestiary():
    kh2 = utils.kh2
    jbtabs = dcc.Tabs(id="JournalBestiaryTabs", value="The Heartless")
    jbtabs.children = [
        dcc.Tab(label="The Heartless", value="The Heartless"),
        dcc.Tab(label="The Nobodies", value="The Nobodies"),
        dcc.Tab(label="Reaction Commands", value="Reaction Commands"),
        dcc.Tab(label="Combo Attacks" if not kh2.fm else "Limits", value="Limits"),
    ]
    return html.Div([
        jbtabs,
        html.Div(id="JournalBestiaryDiv"),
    ])

@callback(
    Input({"type": "Heartless Kill Count", "index": ALL}, "value"),
    State({"type": "Heartless Kill Count", "index": ALL}, "id"),
)
def the_heartless_callback(values, ids):
    kh2 = utils.kh2
    i = 0
    for id in ids:
        idx = id["index"]
        if values[i] is not None:
            kh2.heartless[idx] = values[i]
        i += 1

@callback(
    Input({"type": "Nobody Kill Count", "index": ALL}, "value"),
    State({"type": "Nobody Kill Count", "index": ALL}, "id"),
)
def the_nobodies_callback(values, ids):
    kh2 = utils.kh2
    i = 0
    for id in ids:
        idx = id["index"]
        if values[i] is not None:
            kh2.nobodies[idx] = values[i]
        i += 1

@callback(
    Input({"type": "RC Count", "index": ALL}, "value"),
    State({"type": "RC Count", "index": ALL}, "id"),
)
def rcs_callback(values, ids):
    kh2 = utils.kh2
    i = 0
    for id in ids:
        idx = id["index"]
        if values[i] is not None:
            kh2.rc_usage[idx] = values[i]
        i += 1

@callback(
    Input({"type": "Limit Count", "index": ALL}, "value"),
    State({"type": "Limit Count", "index": ALL}, "id"),
)
def limits_callback(values, ids):
    kh2 = utils.kh2
    i = 0
    for id in ids:
        idx = id["index"]
        if values[i] is not None:
            kh2.limit_usage[idx] = values[i]
        i += 1

File: kh2_src/tabs/journal_tabs/bestiary.js
export function create_bestiary() {
    const kh2jdiv = document.getElementById("kh2jdiv");
    kh2jdiv.innerHTML = `
    <div>
        <div>
            <h3>Sub Entry:</h3>
            <select id="kh2jbtabs">
                <option value="The Heartless">The Heartless</option>
                <option value="The Nobodies">The Nobodies</option>
                <option value="Reaction Commands">Reaction Commands</option>
                <option value="Limits">${window.kh2.fm ? "Limits" : "Combo Attacks"}</option>
            </select>
        </div>
        <div id="kh2jbdiv"></div>
    </div>`;
    const kh2jbtabs = document.getElementById("kh2jbtabs");
    function tab_sel() {
        switch (kh2jbtabs.value) {
            case "The Heartless": {
                create_heartless();
                break;
            }
            case "The Nobodies": {
                create_nobodies();
                break;
            }
            case "Reaction Commands": {
                create_rcs();
                break;
            }
            case "Limits": {
                create_limits();
                break;
            }
        }
    }
    kh2jbtabs.addEventListener("change", tab_sel);
    tab_sel();
}

function create_heartless() {
    const kh2jbdiv = document.getElementById("kh2jbdiv");
    const heartless = `
    <div id="heartless">
        ${
            window.kh2.heartless_list
                .filter((k) => window.kh2.heartless_dict[k] < window.kh2.heartless.length)
                .map((label) => `
                <label style="display: flex; gap: 10px; align-items: center">
                    ${label}:
                    <input
                        type="number"
                        name=${window.kh2.heartless_dict[label]}
                        min=0
                        max=999999
                        step=1
                        value=${window.kh2.heartless[window.kh2.heartless_dict[label]]}
                    >
                </label>`)
                .join("")
        }
    </div>`;
    kh2jbdiv.innerHTML = `
    <div>
        <h3>Heartless Kill Counts:</h3>
        ${heartless}
    </div>`;
    heartless_callbacks();
}

function heartless_callbacks() {
    const heartless = document.getElementById("heartless");
    heartless.addEventListener("change", (e) => {
        if (e.target.validity.valid)
            window.kh2.heartless[e.target.name] = e.target.value;
        e.target.value = window.kh2.heartless[e.target.name];
    });
}

function create_nobodies() {
    const kh2jbdiv = document.getElementById("kh2jbdiv");
    const nobodies = `
    <div id="nobodies">
        ${
            window.kh2.nobody_list
                .map((label) => `
                <label style="display: flex; gap: 10px; align-items: center">
                    ${label}:
                    <input
                        type="number"
                        name=${window.kh2.nobody_dict[label]}
                        min=0
                        max=999999
                        step=1
                        value=${window.kh2.nobodies[window.kh2.nobody_dict[label]]}
                    >
                </label>`)
                .join("")
        }
    </div>`;
    kh2jbdiv.innerHTML = `
    <div>
        <h3>Nobody Kill Counts:</h3>
        ${nobodies}
    </div>`;
    nobodies_callbacks();
}

function nobodies_callbacks() {
    const nobodies = document.getElementById("nobodies");
    nobodies.addEventListener("change", (e) => {
        if (e.target.validity.valid)
            window.kh2.nobodies[e.target.name] = e.target.value;
        e.target.value = window.kh2.nobodies[e.target.name];
    });
}

function create_rcs() {
    const kh2jbdiv = document.getElementById("kh2jbdiv");
    const rcs = `
    <div id="rcs">
        ${
            Object.entries(window.kh2.rc_list_dict)
                .filter(([label, value]) => window.kh2.rc_dict[value[0]] < window.kh2.rc_usage.length)
                .map(([label, value]) => `
                <div style="-webkit-text-size-adjust: 100%; text-size-adjust: 100%;">
                    <h3>${label}:</h3>
                    ${
                        value
                            .map((rc) => `
                            <label style="display: flex; gap: 10px; align-items: center">
                                ${rc}:
                                <input
                                    type="number"
                                    name=${window.kh2.rc_dict[rc]}
                                    min=0
                                    max=9999
                                    step=1
                                    value=${window.kh2.rc_usage[window.kh2.rc_dict[rc]]}
                                >
                            </label>`)
                            .join("")
                    }
                </div>`)
                .join("")
        }
    </div>`;
    kh2jbdiv.innerHTML = `
    <div>
        ${rcs}
    </div>`;
    rcs_callbacks();
}

function rcs_callbacks() {
    const rcs = document.getElementById("rcs");
    rcs.addEventListener("change", (e) => {
        if (e.target.validity.valid)
            window.kh2.rc_usage[e.target.name] = e.target.value;
        e.target.value = window.kh2.rc_usage[e.target.name];
    });
}

function create_limits() {
    const kh2jbdiv = document.getElementById("kh2jbdiv");
    const limit_usage = `
    <div id="limit_usage">
        ${
            window.kh2.limit_list
                .filter((k) => window.kh2.limit_dict[k] < window.kh2.limit_usage.length)
                .map((label) => `
                <label style="display: flex; gap: 10px; align-items: center">
                    ${label}:
                    <input
                        type="number"
                        name=${window.kh2.limit_dict[label]}
                        min=0
                        max=999999
                        step=1
                        value=${window.kh2.limit_usage[window.kh2.limit_dict[label]]}
                    >
                </label>`)
                .join("")
        }
    </div>`;
    kh2jbdiv.innerHTML = `
    <div>
        <h3>Limit Hit Counts:</h3>
        ${limit_usage}
    </div>`;
    limit_usage_callbacks();
}

function limit_usage_callbacks() {
    const limit_usage = document.getElementById("limit_usage");
    limit_usage.addEventListener("change", (e) => {
        if (e.target.validity.valid)
            window.kh2.limit_usage[e.target.name] = e.target.value;
        e.target.value = window.kh2.limit_usage[e.target.name];
    });
}
